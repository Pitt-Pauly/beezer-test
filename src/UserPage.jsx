// @flow
import React, { Component } from 'react';
import styled from 'styled-components';
import firebase from './firebase';
import 'firebase/database';
import spinner from './Facebook-1.1s-134px.svg';
import UserTable from './UserTable';

const Wrapper = styled.div`
  width: 90%;
  margin-top: 25px;
  margin-right: auto;
  margin-bottom: 25px;
  margin-left: auto;
`;

const MessageWrapper = styled.div`
  display: flex;
  align-items: center;
  margin-left: auto;
  margin-right: auto;
  padding-left: 20px;
  padding-right: 20px;
`;

const Spinner = styled.img`
  height: 2.5em;
`;

class UserPage extends Component<
  void,
  {
    apiData: { accounts: Object, users: Object },
    initComplete: boolean,
    error: boolean,
    errorInstance: Object
  }
> {
  state = {
    apiData: {
      accounts: {},
      users: {}
    },
    initComplete: false,
    error: false,
    errorInstance: Object
  };

  componentWillMount() {
    firebase
      .database()
      .ref('/')
      .once('value')
      .then(snapshot => {
        this.setState({
          apiData: snapshot.val(),
          initComplete: true
        });
      })
      .catch(error => {
        this.setState({
          initComplete: false,
          error: true,
          errorInstance: error
        });
      });
  }

  render() {
    const { apiData, initComplete, error, errorInstance } = this.state;

    let content = (
      <MessageWrapper>
        <h2> Loading user data </h2>
        <Spinner src={spinner} className="Spinner" alt="loading spinner" />
      </MessageWrapper>
    );

    if (error === true) {
      content = (
        <MessageWrapper>
          <h2>
            An error occured while fetching the records. Please try again by
            refreshing this page!
          </h2>
          <p>{JSON.stringify(errorInstance)}</p>
        </MessageWrapper>
      );
      return content;
    }

    if (initComplete) {
      content = (
        <Wrapper className="UserTable">
          <UserTable {...apiData} />
        </Wrapper>
      );
    }

    return content;
  }
}

export default UserPage;
