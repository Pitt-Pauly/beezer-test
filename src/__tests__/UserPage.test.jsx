import React from 'react';
import { shallow } from 'enzyme';
import UserPage from '../UserPage';

test('UserPage renders correcty', () => {
  const component = shallow(<UserPage />);
  expect(component).toMatchSnapshot();
});
