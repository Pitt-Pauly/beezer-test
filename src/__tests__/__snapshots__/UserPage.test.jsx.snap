// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPage renders correcty 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserPage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <h2>
           Loading user data 
</h2>,
        <styled.img
          alt="loading spinner"
          className="Spinner"
          src="Facebook-1.1s-134px.svg"
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": " Loading user data ",
        },
        "ref": null,
        "rendered": " Loading user data ",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "alt": "loading spinner",
          "className": "Spinner",
          "src": "Facebook-1.1s-134px.svg",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <h2>
             Loading user data 
</h2>,
          <styled.img
            alt="loading spinner"
            className="Spinner"
            src="Facebook-1.1s-134px.svg"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": " Loading user data ",
          },
          "ref": null,
          "rendered": " Loading user data ",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "alt": "loading spinner",
            "className": "Spinner",
            "src": "Facebook-1.1s-134px.svg",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
